timemachine/cpp/src/barostat.cu

```diff
@@ -0,0 +1,267 @@
+#include "barostat.hpp"
+#include "gpu_utils.cuh"
+#include "fixed_point.hpp"
+#include <algorithm>
+#include <stdio.h>
+
+#define AVOGADRO 6.0221367e23
+#define BOLTZ 0.008314462618
+
+namespace timemachine {
+
+MonteCarloBarostat::MonteCarloBarostat(
+    const int N,
+    const double pressure,  // Expected in Bar
+    const double temperature, // Kelvin
+    const std::vector<std::vector<int> > group_idxs,
+    const int interval,
+    const std::vector<BoundPotential *> bps,
+    const int seed) :
+    N_(N),
+    pressure_(pressure),
+    temperature_(temperature),
+    interval_(interval),
+    bps_(bps),
+    group_idxs_(group_idxs),
+    volume_scale_(0),
+    num_attempted_(0),
+    num_accepted_(0),
+    seed_(seed),
+    step_(0) {
+
+    // lets not have another facepalm moment again...
+    if(temperature < 100.0) {
+        std::cout << "warning temperature less than 100K" << std::endl;
+    }
+
+    if(pressure > 10.0) {
+        std::cout << "warning pressure more than 10bar" << std::endl;
+    }
+
+    mt_ = std::mt19937(seed_);
+    dist_ = std::uniform_real_distribution<double>(0.0, 1.0);
+
+    const int num_mols = group_idxs_.size();
+
+    gpuErrchk(cudaMalloc(&d_x_after_, N_*3*sizeof(*d_x_after_)));
+    gpuErrchk(cudaMalloc(&d_box_after_, 3*3*sizeof(*d_box_after_)));
+
+    gpuErrchk(cudaMalloc(&d_u_before_, N_*sizeof(*d_u_before_)));
+    gpuErrchk(cudaMalloc(&d_u_after_, N_*sizeof(*d_u_after_)));
+
+    gpuErrchk(cudaMalloc(&d_mol_idxs_, num_mols*sizeof(*d_mol_idxs_)));
+
+    int mol_idxs[num_mols];
+    for (int i = 0; i < num_mols; i++) {
+        std::vector<int> atoms = group_idxs[i];
+        // Sort the array in ascending order
```

Comment:
Why does each mol need have its atoms be sorted? To help with faster mem loading?

Reply:
Because the group indices are not always ordered so you could get something like...

```
[[1,2,3], [6,5,4]]
```
which would result in the logic thinking that the atom is of size 5 rather than 3. Could also find the min of the next mol, which would be more consistent in its run time.